### PR TITLE
Fix tests crashing when accessing SPM dependency resource bundle

### PR DESCRIPTION
**Pull request by rhysm94**
Resolves <https://github.com/tuist/tuist/issues/6885>

### Short description 📝

This fixes the issue of tests crashing when attempting to access a resource that is in an SPM dependency's resource bundle.

The main issue is that when testing a module that depends on another module which depends on an SPM-imported dependency with a resource, none of the URLs used in the lookup of that SPM module's bundle work.
`Bundle.main` appears to refer to the Xcode test agent, and the bundles for the SPM dependency are not copied into the framework, so `Bundle(for: BundleFinder.self).resourceURL` also won't find it. This leads to a `fatalError` because no bundle can be found.

I'm presuming, based on other PR titles, that it's intentional that the `*.bundle` files are no longer copied into the framework for the module being tested. Therefore, I've solved this by adding a new lookup URL in the `Bundle.module` accessor. Here, I'm using `Bundle(for: BundleFinder.self).resourceURL.appendingPathComponent("..")` to look in the parent directory, which *does* appear to contain the appropriate `*.bundle` files for the imported SPM dependency.

### How to test the changes locally 🧐

I've included a new minimum-replicable example project called `framework_with_spm_bundle` that imports the Stripe SDK, as this is the initial framework I encountered the issue with! More than happy to switch this out for another SPM bundle that replicates the same idea, including creating a smaller example of my own, if you'd prefer?

I've added a new test that runs the tests for `framework_with_spm_bundle`. This previously crashed before I added the changes in this PR, and can be tested by checking out the second commit then checking out the most recent commit to confirm it's fixed.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Review by fortmarek — COMMENTED** [review]
hey 👋 

Thanks a lot for the PR – resources are tricky 😅 

I recently fixed a similar case but for previews: https://github.com/tuist/tuist/pull/6865

Would you mind trying to revert these [changes](https://github.com/tuist/tuist/pull/6865/files#diff-e92bb6ba816dd394aacf7d74e51dc2269fc21f4b2f1f27031ffede91837cb0e5) and seeing if the previews in the `app_with_previews` still work? If they do, we can go ahead with your changes and revert the changes I did in the previous PR.

**Comment by rhysm94** [comment]
> hey 👋
> 
> Thanks a lot for the PR – resources are tricky 😅
> 
> I recently fixed a similar case but for previews: #6865
> 
> Would you mind trying to revert these [changes](https://github.com/tuist/tuist/pull/6865/files#diff-e92bb6ba816dd394aacf7d74e51dc2269fc21f4b2f1f27031ffede91837cb0e5) and seeing if the previews in the `app_with_previews` still work? If they do, we can go ahead with your changes and revert the changes I did in the previous PR.

Hey, thanks! I'm having some slight issues running tuist using the `swift run` command. I keep getting errors about `could not build Objective-C module 'ProjectDescription'` even though I've run `swift build --product ProjectDescription` before:
e.g.
```sh
/Users/rhysmorgan/Developer/tuist/Tuist/Config.swift:1:8: error: could not build Objective-C module 'ProjectDescription'
 1 | import ProjectDescription
   |        `- error: could not build Objective-C module 'ProjectDescription'
 2 | 
 3 | let config = Config(
```

Could you provide any guidance on how I could perhaps address this, so I can test if reverting that change works with Previews too? Thanks!

**Comment by fortmarek** [comment]
@rhysm94 that's a known issue flagged [here](https://github.com/tuist/tuist/issues/6775) that started to happen from Xcode 16. We haven't had time to look into that, yet.

We recommend running Tuist via Xcode instead of the CLI: https://docs.tuist.io/en/contributors/get-started#from-xcode

Let me know if that works for you :)

**Comment by rhysm94** [comment]
@fortmarek - Reverting your change in `GraphTraverser`, i.e. setting it back to:

```swift
isDependencyResourceBundle(dependency: dependency) &&
                (isDependencyExternal(dependency) || dependency.isPrecompiled) &&
                canEmbedBundles(target: target)
```

on top of my changes, and generating that project all seems to work!

When I open the `TestView` in PreviewsFramework, and click "Click to read file from bundle", the text "hello" from the txt file appears underneath without crashing.

**Comment by fortmarek** [comment]
>  Reverting your change in GraphTraverser, i.e. setting it back to on top of my changes, and generating that project all seems to work!

Actually, scratch that, I don't think that will work in the following scenario:
```
tuist cache
tuist generate
```

And then if you try to run the `PreviewsFramework` preview, the bundle would not be in the derived data at all. Let's keep that change in.

Let's focus on the case you set out to fix initially (my bad!).

@rhysm94, can you try replicating the same graph but with using the vanilla Xcode integration (i.e., depending on `.package(...)` instead of `.external(...)`? The current bundle accessor is taken from there, so I wonder why we need to diverge here. It'd be primarily interesting to compare how the derived data looks in their case compared to ours.

**Comment by rhysm94** [comment]
@fortmarek OK, will revert that previous commit and let you know how they differ.

**Comment by rhysm94** [comment]
@fortmarek Sorry if I'm missing something, but when I change `.external(...)` to `.package(...)` in my test repo - https://github.com/rhysm94/TuistBundleBug/blob/main/Project.swift - I get "Missing package product 'Stripe'" when attempting to build any of the modules.

Is there anything else I need to change? I've run `tuist install` before re-running `tuist generate`, but that doesn't seem to help.

**Comment by fortmarek** [comment]
> I get "Missing package product 'Stripe'" when attempting to build any of the modules.

You also need to add `Stripe` to the list of packages in `Project.swift` directly, not in `Package.swift`.

**Comment by rhysm94** [comment]
@fortmarek Not sure what you mean by that, sorry!
Do I need to add Stripe as a target? e.g.

```swift
targets: [
  .target(
    name: "Stripe",
    ...
  ),
  .target(
    name: "SharedUI",
    destinations: .iOS,
    product: .framework,
    bundleId: "co.rhysmorgan.TuistTest.sharedUI",
    sources: ["SharedUI/Sources/**"],
    dependencies: [
      .package(product: "Stripe")
    ]
  ),
]
```

or is there some other change I need to make here?

**Comment by fortmarek** [comment]
You need to add the package also as a `Project.packages` property:
```swift
let project = Project(name: "MyProject", packages: [
    .remote(url: "https://github.com/krzyzanowskim/CryptoSwift", requirement: .exact("1.8.0"))
])
```

See the docs [here](https://docs.tuist.io/en/guides/develop/projects/dependencies), particularly the `Xcode's default integration` section.

**Comment by rhysm94** [comment]
> You need to add the package also as a `Project.packages` property:
> 
> ```swift
> let project = Project(name: "MyProject", packages: [
>     .remote(url: "https://github.com/krzyzanowskim/CryptoSwift", requirement: .exact("1.8.0"))
> ])
> ```
> 
> See the docs [here](https://docs.tuist.io/en/guides/develop/projects/dependencies), particularly the `Xcode's default integration` section.

Ahhh, thank you!

OK, I've made that change, adding it in `Project.swift` instead, e.g.:

```swift
let project = Project(
  name: "TuistTest",
  packages: [
    .remote(url: "https://github.com/stripe/stripe-ios-spm", requirement: .exact("23.27.3")),
    .remote(url: "https://github.com/pointfreeco/swift-snapshot-testing", requirement: .exact("1.17.5")),
  ],
```

and removing it from `Package.swift`.
When I run the tests, it crashes with the exact same error: 

`StripePaymentsUI/resource_bundle_accessor.swift:44: Fatal error: unable to find bundle named Stripe_StripePaymentsUI`

**Comment by fortmarek** [comment]
Hey @rhysm94,

using a pure SPM project was not what I had in mind. I wanted to test how it would work if you had the same graph as in your initial reproduction but instead of using the XcodeProj-based integration, you would use the vanilla one.

I created such a repro sample here: https://github.com/fortmarek/TuistBundleBug

The tests still crash in that case, so this is not really a Tuist bug, but an SPM bug.

The issue is that SPM aggressively copies the bundle to the downstream app target, which means the `.bundle` is not part of the `.framework` when the framework is run in isolation.

We could go with your fix, but I'm not totally convinced that we want to rely that heavily on the inner structure of the derived data and add additional workarounds on top of what is already quite smelly SPM code.

What you can do instead to get the App main bundle in your tests is to add `.target(name: "TuistTest"),` to the dependencies of `PaymentsTests`. That will set up `TuistTest` as the host app and the bundle locator will find the Stripe Payments bundle as it will be part of the main bundle of the app.

**Comment by rhysm94** [comment]
@fortmarek Apologies for poorly explaining. I tried that option too, and saw exactly what you've also seen – that using the `Project.swift` file for declaring the remote dependencies still caused the crash. I forgot to push my change here to demonstrate 🤦🏻 https://github.com/rhysm94/TuistBundleBug

I created the pure SPM example just to see how it compared to the Tuist-generated version.

Adding `.target(name: "TuistTest")` does indeed fix things!
Do you know if there any further implications for linking my test modules against the app target?

**Comment by rhysm94** [comment]
@fortmarek – I've restructured my reproduction example at https://github.com/rhysm94/TuistBundleBug to be more like the "real" work project (and the fixture I included in this PR), and now the solution of linking `PaymentsTests` against `TuistTest` no longer resolves the problem. (Quite glad actually, as I was seeing this in the real project, and getting worried that I wouldn't be able to resolve this issue!)

When I run `twist generate` from this branch, however, the tests do still pass.

Completely understand that it's a potentially brittle solution, reliant on the structure of a folder that none of us control, but it does still seem to be _a_ solution. If it's possible to be aggressive in copying bundles in when building Unit Tests, without having an impact on real apps, I'd be happy to give that a shot instead though?

**Comment by fortmarek** [comment]
> I've restructured my reproduction example at https://github.com/rhysm94/TuistBundleBug to be more like the "real" work project (and the fixture I included in this PR), and now the solution of linking PaymentsTests against TuistTest no longer resolves the problem. (Quite glad actually, as I was seeing this in the real project, and getting worried that I wouldn't be able to resolve this issue!)

I tried the sample project but it doesn't build. The bundle should _always_ be there when you use a host app and when I updated the manifest to again have a host app (which your latest commit does not), the tests do still work.

Would you mind cleaning up the sample and providing exact steps that I can reproduce?

**Comment by rhysm94** [comment]
> I tried the sample project but it doesn't build. The bundle should _always_ be there when you use a host app and when I updated the manifest to again have a host app (which your latest commit does not), the tests do still work.
> 
> Would you mind cleaning up the sample and providing exact steps that I can reproduce?

Apologies, I don't know how I didn't push the latest version of my branch. 🤦🏻‍♂️ Last week was a long week...
It now consists of directories with their own Project.swift files, which is how the "real world" app is structured.

When going *back* to the *main* branch, I've attempted to add the host app as a target to the Payments module in a couple of different ways.
If I add `.target(name: "TuistTest")` or `.project(target: "TuistTest", path: "..")` to the PaymentTests' dependencies array, I get this completely expected error:

```
[31mCouldn't find target 'TuistTest' at '/Users/rhysmorgan/Developer/TuistTest/Payments'
Consider creating an issue using the following link: https://github.com/tuist/tuist/issues/new/choose[0m
```

If I add `.project(target: "TuistTest", path: "..")` instead, then Tuist generates the Xcode workspace. When I attempt to run `testMyViewSnapshot` in PaymentsTests, the test crashes with the same `StripePaymentsUI/TuistBundle+StripePaymentsUI.swift:37: Fatal error: unable to find bundle named Stripe_StripePaymentsUI` as before.

Switching back to this `fix-missing-spm-bundle` branch, running `tuist generate` on my sample project (regardless of whether I've included the extra project dependency), and attempting to run that same test succeeds without issue.

Full repro:

1. Pull current main from https://github.com/rhysm94/TuistBundleBug
2. Amend Payments/Project.swift to add `.project(target: "TuistTest", path: "..")` as a dependency of the PaymentsTests target
3. Run `tuist generate` on the TuistBundleBug directory.
4. Attempt to run the `MyViewSnapshotTests` in Payments > Tests, and see `Fatal error` from above.
6. Remove the change in Payments/Project.swift
7. Switch to `fix-missing-spm-bundle` branch
8. Run `tuist generate` on the TuistBundleBug directory again.
9. Run the `MyViewSnapshotTests` successfully

**Comment by fortmarek** [comment]
Gotcha. I believe the problem is we currently don't allow specifying a test host app across different projects. There's a PR to fix this: https://github.com/tuist/tuist/pull/6820

After that's merged, the tests should pass again. If you look at the generated project, the `Payments` unit tests target should be missing a reference to the test host app (which is a bug in Tuist).

**Comment by rhysm94** [comment]
> Gotcha. I believe the problem is we currently don't allow specifying a test host app across different projects. There's a PR to fix this: #6820
> 
> After that's merged, the tests should pass again. If you look at the generated project, the `Payments` unit tests target should be missing a reference to the test host app (which is a bug in Tuist).

Yep, there's no evidence in the Payments unit test target that it's linked against the test host app

**Comment by fortmarek** [comment]
> Yep, there's no evidence in the Payments unit test target that it's linked against the test host app

You can run the sample from the linked PR's branch to see if things work for you with the fix there.

**Comment by rhysm94** [comment]
> You can run the sample from the linked PR's branch to see if things work for you with the fix there.

OK, I've done that.
1. Re-added the `.project(target: "TuistTest", path: "..")` line
2. Cloned the `testhost-non-local` branch, and ran `tuist generate` from it on the `TuistBundleBug` dir.  
  I can see that it is indeed now included in the "Frameworks and Libraries" section!
3. I've cleaned and rebuilt the `TuistBundleBug` project, and attempted to run the `MyViewSnapshotTests`
4. Unfortunately, I get the same crash as before.

Screenshot 2024-10-28 at 09 08 53

**Comment by fortmarek** [comment]
>  I can see that it is indeed now included in the "Frameworks and Libraries" section!

But the test host app is still not set. If you manually set the test host app to `TuistTest`, the tests should pass.

@TamarMilchtaich, shouldn't we try to set the host app from non-local dependencies [here](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift#L288)?

**Comment by rhysm94** [comment]
> >  I can see that it is indeed now included in the "Frameworks and Libraries" section!
> 
> 
> 
> But the test host app is still not set. If you manually set the test host app to `TuistTest`, the tests should pass.
> 
> 
> 
> @TamarMilchtaich, shouldn't we try to set the host app from non-local dependencies [here](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift#L288)?

Do you mean set the scheme to TuistTest? Or is there some other setting that I've missed?

**Comment by fortmarek** [comment]
> Do you mean set the scheme to TuistTest? Or is there some other setting that I've missed?

In the screenshot, the `Host Application` is set to `None`

**Comment by fortmarek** [comment]
> Sorry, noticed that just now. It doesn't appear as an option in that list

Got it. Either way, with that PR, you should be able to set a test host app from a different project, so this PR should no longer be necessary. As mentioned, we could still do it, but I'd rather not diverge from SPM's behavior.

**Comment by rhysm94** [comment]
> > Sorry, noticed that just now. It doesn't appear as an option in that list
> 
> Got it. Either way, with that PR, you should be able to set a test host app from a different project, so this PR should no longer be necessary. As mentioned, we could still do it, but I'd rather not diverge from SPM's behavior.

I'm a little confused as to setting the test host app "from a different project"? Is that not what I've just tried, using the `testhost-non-local` PR, and it hasn't worked? It's not appearing in the list, and I can't run the tests without crashing.

I get not liking this particular fix, but it does solve the problem that exists since a breaking change in Tuist 4.22.0. I'm completely open to trying to solve the problem in a different way, one that means I can run my feature module tests in isolation still, but I don't really know where to begin with that. Is there a way to aggressively copy in bundles when building unit tests, and retain the current not-copying behaviour when building an app target? Could a reasonable solution be to include the fix in this PR while I try and figure out the above?

**Comment by fortmarek** [comment]
> I'm a little confused as to setting the test host app "from a different project"? Is that not what I've just tried, using the testhost-non-local PR, and it hasn't worked? It's not appearing in the list, and I can't run the tests without crashing.

There's a bug in that PR which is work-in-progress. You should be able to point to a host application from a different project, afaik. That's my preferred solution, rather than somehow copying things around.

> Could a reasonable solution be to include the fix in this PR while I try and figure out the above?

I'd rather focus on merging the `testhost-non-local` PR. The issue I have here is that on a device, we can look for the bundle in directories that don't belong to us (and again, diverging from SPM).

**Comment by rhysm94** [comment]
> There's a bug in that PR which is work-in-progress. You should be able to point to a host application from a different project, afaik. That's my preferred solution, rather than somehow copying things around.

Does that mean I'd have to build the entire application just to test a feature module though? If so, that kinda defeats the biggest benefits of modularising a codebase versus a monolith 😅

I know that this PR's change diverges from how SPM does things slightly, but does Tuist not diverge from SPM's bundle copying behaviour?  
It's a pretty small divergence too, just giving one extra directory for the `Bundle.module` lookup to try.

**Comment by TamarMilchtaich** [comment]
> I'm a little confused as to setting the test host app "from a different project"? Is that not what I've just tried, using the testhost-non-local PR, and it hasn't worked? It's not appearing in the list, and I can't run the tests without crashing.

There's no need to choose the test host app from the list, you can see an example here:
https://tuist-community.slack.com/archives/C018QG7U7SN/p1722348832047419?thread_ts=1722348710.616589&cid=C018QG7U7SN
Which works with `testhost-non-local` with nothing selected there.

Note that adding the test host as a dependency is not enough, one also needs to add the relevant settings:
```
    "BUNDLE_LOADER": "$(BUILT_PRODUCTS_DIR)/TuistTest.app/TuistTest",
    "TEST_HOST": "$(BUNDLE_LOADER)",
```
(or adjust the bundle loader according to the actual path)

Does it work when adding those settings?

**Comment by TamarMilchtaich** [comment]
> @TamarMilchtaich, shouldn't we try to set the host app from non-local dependencies [here](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Settings/DefaultSettingsProvider.swift#L288)?

@fortmarek These are the settings mentioned above, I'll add it to the PR as well

**Comment by TamarMilchtaich** [comment]
> Could you explain what's on the Slack message please?

It merely demonstrates a project with the settings from above and a non-local test host, so nothing new, just thought an example might help

>  I get an error about a cyclical dependency 

Does the error persist if you clean the derived data? I think it happened to me after switching to similar configuration. If the TuistTest target doesn't really depend on `PaymentsTests`, it makes sense that it's not a real issue, but something residual in the build folder. Does it persist after cleaning the derived data?

**Comment by fortmarek** [comment]
> If your concern is that this allows you to read a resource from a directory that we don't own, could I resolve this by adding a check that we're in a unit test context before adding the parent directory check, and/or wrapping it in an if #DEBUG guard?

`if #DEBUG` would probably make things worse as things would only start to crash once you build a release variant of your app.

> My concern with the host app solution is that this means we have to build the entire app target just to run tests on a feature module, which feels like a big regression in benefits for modularising a codebase.

I hear you – although we're in this situation due to Xcode/SPM's mistakes, not Tuist's.

Is there a way to tell if we're running tests/running the framework in isolation and apply the workaround only when one of those conditions is true? I'd love to apply the workaround only when really needed – similarly to the workaround applied for [previews](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Mappers/ResourcesProjectMapper.swift#L266-L269) that's not run outside of the context of previews.

**Comment by rhysm94** [comment]
> `if #DEBUG` would probably make things worse as things would only start to crash once you build a release variant of your app.

This would only apply if running unit tests against a release variant of a framework though, I think, right? Is this something that’s likely to happen?
Maybe an alternative could be something like `#if canImport(XCTest)` as a guard, maybe?

> I hear you – although we're in this situation due to Xcode/SPM's mistakes, not Tuist's.
> 
> Is there a way to tell if we're running tests/running the framework in isolation and apply the workaround only when one of those conditions is true? I'd love to apply the workaround only when really needed – similarly to the workaround applied for [previews](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Mappers/ResourcesProjectMapper.swift#L266-L269) that's not run outside of the context of previews.

Agreed, it’d definitely be preferable to apply the workaround only in a unit test situation. I’ll look into the `ResourceProjectMapper` file to see if there’s any way to apply a change there instead.

**Comment by fortmarek** [comment]
>  Maybe an alternative could be something like #if canImport(XCTest) as a guard, maybe?

Yes, if that works, that'd be perfect!

**Comment by rhysm94** [comment]
> Yes, if that works, that'd be perfect!

Just tested it locally, and it appears to work as a solution, gating appending the parent directory lookup behind `canImport(XCTest)`!

**Review by fortmarek — COMMENTED** [review]
Thanks! I feel better about the change. Two minor comments after which we should be good to merge 🙂

File: Sources/TuistAcceptanceTesting/TuistAcceptanceFixtures.swift [hunk 1]
```diff
@@ -24,6 +24,7 @@ public enum TuistAcceptanceFixtures {
     case frameworkWithMacroAndPluginPackages
     case frameworkWithNativeSwiftMacro
     case frameworkWithSwiftMacro
+    case frameworkWithSPMBundle
```
[the remark this answers is not in the document]
Comment (rhysm94) [reply Sources/TuistAcceptanceTesting/TuistAcceptanceFixtures.swift:27]: I've already added that one here [`test_with_framework_with_spm_bundle`](https://github.com/rhysm94/tuist/blob/729a068533bc05a0fccae84aee5be05395157f67/Tests/TuistAutomationAcceptanceTests/TestAcceptanceTests.swift#L24)
😀

[the remark this answers is not in the document]
Comment (fortmarek) [reply Sources/TuistAcceptanceTesting/TuistAcceptanceFixtures.swift:27]: Ahh, sorry, I missed that!

**Review by fortmarek — APPROVED** [review]
LGTM 🚀

**Comment by fortmarek** [comment]
@all-contributors add @rhysm94 for code

**Comment by Gladkov-Art** [comment]
Hi! I am still having the same issue on **4.34.3**. I'm having an Objective-C accessor generated for SPM dependency. At runtime it looks like this:

Screenshot 2024-11-17 at 17 09 50

**Pull request merged**